File: 프로그래머스/3/60059. 자물쇠와 열쇠/자물쇠와 열쇠.py
def rotate(key, n, m):
    arr = [[0] * m for i in range(m)]
    
    for i in range(m):
        for j in range(m):
            arr[j][m-1-i] = key[i][j]
    
    return arr


def check(x, y, key, lock, n, m) -> bool:
    lock_tmp = [[0] * n for i in range(n)]
    
    # 복제
    for i in range(n):
        for j in range(n):
            lock_tmp[i][j] = lock[i][j]
            
    # key 옮기기
    for i in range(m):
        for j in range(m):
            if 0<=x+i<n and 0<=y+j<n:
                if key[i][j]^lock_tmp[x+i][y+j] == 0: return False
                lock_tmp[x+i][y+j] = 1
    
    # 합이 0 맞는지 확인하기
    # return sum(lock_tmp) == n * n
    return all(all(cell == 1 for cell in row) for row in lock_tmp)


def solution(key, lock):
    answer = False
    
    # key 시작 위치
    m = len(key)
    n = len(lock)
    
    for dir in range(0, 4):
        for x in range(-n, n):
            for y in range(-n, n):
                answer |= check(x, y, key, lock, n, m)
        key = rotate(key, n, m)
            
            
    return answer

File: 프로그래머스/3/60059. 자물쇠와 열쇠/자물쇠와 열쇠.cpp
#include <string>
#include <vector>
#include <iostream>
using namespace std;

int n, m;

// 시계방향 90도 회전
void rotate(vector<vector<int>>& key){
    vector<vector<int>> tmp(m, vector<int>(m, 0));
    
    for(int i=0; i<m; i++){
        for(int j=0; j<m; j++){
            tmp[j][m - 1 - i] = key[i][j];
        }
    }
    
    key.swap(tmp);
    
}

bool in_range(int x, int y){
    return 0<=x && 0<=y && x <n && y < n;
}

bool solution(vector<vector<int>> key, vector<vector<int>> lock) {
    bool answer = true;
    m = key.size();
    n = lock.size();
    
    for(int dir=0; dir<4; dir++){
        
        // key 시작 좌표
        for(int i=-n; i<n; i++){
            for(int j=-n; j<n; j++){
                vector<vector<int>> lock_tmp = lock;
                
                bool is_pos = true; // 가능한지 여부
                
                // key 옮기기
                for(int x=0; x<m; x++){
                    for(int y=0; y<m; y++){
                        if(!in_range(i+x, j+y)) continue; // 범위 나간 경우
                        // 어차피 안되는 경우
                        if(lock_tmp[i+x][j+y] == 1 && key[x][y] == 1) { is_pos = false; break;} 
                        if(lock_tmp[i+x][j+y] == 0 && key[x][y] == 0) { is_pos = false; break;}
                        lock_tmp[i+x][j+y] = 1;
                    }
                }
                
                if(is_pos == false) continue; // 해당 경우 불가능
                
                for(int x=0; x<n; x++){
                    for(int y=0; y<n; y++){
                        if(lock_tmp[x][y] == 0) is_pos = false;
                    }
                }
                
                if(is_pos == false) continue;
                return true;
            }
        }
        
        rotate(key);

    }

    
    return false;
}